### ISSUE TITLE
lua install steps

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bcc#512]
Either in INSTALL.md, or in docs/lua.md and linked from INSTALL.md. ... Might want a docs/python.md as well at some point.

I haven't figured it out yet. Here's what I get when I run an example:

```
/mnt/src/bcc/examples/lua# lua bashreadline.lua 
lua: bashreadline.lua:16: module 'ffi' not found:
	no field package.preload['ffi']
	no file './ffi.lua'
	no file '/usr/local/share/lua/5.1/ffi.lua'
	no file '/usr/local/share/lua/5.1/ffi/init.lua'
	no file '/usr/local/lib/lua/5.1/ffi.lua'
	no file '/usr/local/lib/lua/5.1/ffi/init.lua'
	no file '/usr/share/lua/5.1/ffi.lua'
	no file '/usr/share/lua/5.1/ffi/init.lua'
	no file './ffi.so'
	no file '/usr/local/lib/lua/5.1/ffi.so'
	no file '/usr/lib/x86_64-linux-gnu/lua/5.1/ffi.so'
	no file '/usr/lib/lua/5.1/ffi.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
stack traceback:
	[C]: in function 'require'
	bashreadline.lua:16: in main chunk
	[C]: ?
```

Ok, am I supposed to be running it from lua or luajit?:

```
/mnt/src/bcc/examples/lua# /mnt/LuaJIT-2.0.4/src/luajit bashreadline.lua 
/mnt/src/bcc/examples/lua#
```

I'm probably just doing it wrong. //CC @vmg

**Comment by vmg** [comment]
Hi Brendan! Sorry about the lack of documentation!

The recommended way to run the lua probes is the `bcc-lua` tool. If CMake detects the `luajit` and `luajit-5.1-dev` packages (these are the Ubuntu names), it will create a standalone `bcc-lua` tool with no dependencies that run any of the Lua probes and has an user friendly interface.

You'll find it at `build/src/lua/bcc-lua` (again, if the Lua packages are available). This tool has 0 dependencies (so it can be deployed to servers) and will load `libbcc.so` dynamically at runtime (to let you switch between libbcc versions). Setting the `LIBBCC_SO_PATH` environment variable will let you point the tool at the version of the library you want to be using.

From there it's smooth sailing. Try `./bcc-lua --help` to see the available options.

**Comment by 4ast** [comment]
I think we still need to find a way to statically link libbcc into bcc-lua. That will solve libelf.so dependency too.

**Comment by vmg** [comment]
Oh, that's totally doable. Let me open a PR.

**Comment by brendangregg** [comment]
Thanks, works! ... Mostly works: :)

```
# bcc-lua bashreadline.lua 
TIME      PID    COMMAND
17:50:32  9714   ls
17:50:35  9714   ls -l
17:50:43  9714   echo I am using bcc-lua!
17:50:50  9714   foo bar baz
^Cwrite(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
write(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
write(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
write(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
write(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
write(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
write(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
write(/sys/kernel/debug/tracing/kprobe_events): No such file or directory
```

Maybe I'll whip up a brief lua.md install instructions as a starting point.

I think those lua examples should be chmod a+x, and #!/usr/bin/bcc-lua. Apart from making them easier to run (more like tools), it self documents the dependency: bcc-lua.

**Comment by brendangregg** [comment]
PS, if it helps with those kprobe errors, this was in dmesg:

```
[2686884.549532] Event kprobes/perf_event_array:0:7 doesn't exist.
[2686884.554960] Event kprobes/perf_event_array:0:2 doesn't exist.
[2686884.557844] Event kprobes/perf_event_array:0:6 doesn't exist.
[2686884.560623] Event kprobes/perf_event_array:0:1 doesn't exist.
[2686884.563339] Event kprobes/perf_event_array:0:5 doesn't exist.
[2686884.566030] Event kprobes/perf_event_array:0:0 doesn't exist.
[2686884.568710] Event kprobes/perf_event_array:0:4 doesn't exist.
[2686884.571378] Event kprobes/perf_event_array:0:3 doesn't exist.
```

**Comment by vmg** [comment]
The probe errors when deatching (and a lot more) fixed in #515

**Comment by brendangregg** [comment]
Closing ticket, I think we have instructions now. Still might want a separate docs/lua.md for more details later on.
